include/cantera/kinetics/ReactionRate.h

```diff
@@ -11,473 +11,180 @@
 #include "cantera/base/ctexceptions.h"
 
 namespace Cantera
 {
 
-class Func1;
-class MultiRateBase;
-
-//! Abstract base class for reaction rate definitions
-/**
- * Because this class has no template parameters, derived objects can be
- * accessed via `shared_ptr<ReactionRateBase>`. For performance reasons
- * it is essential that derived classes use the keyword `final` to
- * de-virtualize `virtual` methods.
- *
- * Methods defined for the abstract base class are not aware of specialized
- * data handlers defined by the template class `ReactionRate<DataType>`
- * and thus can be exposed to the API.
- */
-class ReactionRateBase
+//! Abstract base class for reaction rate definitions; this base class is used by
+//! user-facing APIs to access reaction rate objects
+//!
+//! In addition to the pure virtual methods declared in this class, complete derived
+//! classes must implement the method `evalFromStruct(const DataType& shared_data)`,
+//! where `DataType` is a container for parameters needed to evaluate reactions of that
+//! type.
+//!
+//! Derived classes may also implement the methods
+//! `updateFromStruct(const DataType& shared_data)` and
+//! `ddTFromStruct(const DataType& shared_data)`.
+class ReactionRate
 {
 public:
-    ReactionRateBase() : units(0.) {}
-    virtual ~ReactionRateBase() {}
+    ReactionRate() : m_rate_index(npos) {}
+
+    // Copy constructor and assignment operator need to be defined because of the
+    // #m_evaluator member that can't (and shouldn't) be copied.
+    ReactionRate(const ReactionRate& other)
+        : m_input(other.m_input)
+        , m_rate_index(other.m_rate_index)
+    {}
+
+    ReactionRate& operator=(const ReactionRate& other) {
+        if (this == &other) {
+            return *this;
+        }
+        m_input = other.m_input;
+        m_rate_index = other.m_rate_index;
+        return *this;
+    }
+
+    virtual ~ReactionRate() = default;
+
+    //! Create a rate evaluator for reactions of a particular derived type.
+    //! Derived classes usually implement this as:
+    //!
+    //! ```.cpp
+    //! unique_ptr<MultiRateBase> newMultiRate() const override {
+    //!     return unique_ptr<MultiRateBase>(new MultiBulkRate<RateType, DataType>);
+    //! ```
+    //!
+    //! where `RateType` is the derived class name and `DataType` is the corresponding
+    //! container for parameters needed to evaluate reactions of that type.
+    virtual unique_ptr<MultiRateBase> newMultiRate() const = 0;
 
-public:
-    //! Identifier of reaction type
-    virtual std::string type() const = 0;
+    virtual const std::string type() const = 0;
 
-    //! Create multi-rate evaluator
-    virtual unique_ptr<MultiRateBase> newMultiRate() const = 0;
+    //! Set parameters
+    //! @param node  AnyMap object containing reaction rate specification
+    //! @param rate_units  unit definitions specific to rate information
+    virtual void setParameters(const AnyMap& node, const UnitStack& units) {
```

Comment:
The name of the second argument is mismatched between the docstring and the function signature (which will cause warnings from Doxygen)